libraries/ref-extractor.js

```diff
@@ -34,52 +34,75 @@ function handleFormatSelect(event) {
-              if (endRun.length != 0) {
-                  break;
-              }
-              
-              // Concatenate textContents of <w:instrText/> elements within complex field
-              instrTextFields = nextRun.getElementsByTagName("w:instrText");
-              for (let i = 0; i < instrTextFields.length; i++) {
-                  instrTextContent += instrTextFields[i].textContent;
-              }
-              
-              nextRun = nextRun.nextSibling;
+        var parsedDOM = new DOMParser().parseFromString(xmlDocumentFile, 'text/xml');
+
+        if (documentType == "Word") {
+            // Locate the beginning of complex fields (<w:fldChar w:fldCharType="begin"/>, child of <w:r>)
+            var complexFieldStarts = parsedDOM.querySelectorAll("*|fldChar[*|fldCharType=begin]");
+
+            for (var i = 0; i < complexFieldStarts.length; i++) {
+                instrTextContent = "";
+
+                // Visit sibling <w:r> elements until we hit the last one (<w:fldChar w:fldCharType="end"/>, child of <w:r>)
+                nextRun = complexFieldStarts[i].parentElement.nextSibling;
+                while (nextRun) {
+                    endRun = nextRun.querySelectorAll("*|fldChar[*|fldCharType=end]");
+                    if (endRun.length != 0) {
+                        break;
+                    }
+
+                    // Concatenate textContents of <w:instrText/> elements within complex field
+                    instrTextFields = nextRun.getElementsByTagName("w:instrText");
+                    for (let i = 0; i < instrTextFields.length; i++) {
+                        instrTextContent += instrTextFields[i].textContent;
+                    }
+
+                    nextRun = nextRun.nextSibling;
+                }
+                fields.push(instrTextContent);
+            }
+        } else if (documentType == "OpenDocument") {
+            // when using reference marks, Zotero stores citations in the
+            // text:name attribute of <text:reference-mark-start> elements
+            var referenceMarks = parsedDOM.querySelectorAll("*|reference-mark-start[*|name]");
+
+            for (var i = 0; i < referenceMarks.length; i++) {
+                fields.push(referenceMarks[i].getAttribute("text:name"));
             }
-            fields.push(instrTextContent);
         }
         
         return(fields);
     }
 
     JSZip.loadAsync(file).then(function(zip) {
-        // "word/document.xml" seems to contain "author-date" style CSL citations
-        // "word/footnotes.xml" seems to contain "note" style CSL citations
-        filesToExtract = ["word/document.xml", "word/footnotes.xml", "word/endnotes.xml"];
-        
+        var filesToExtract = [];
+        var fileWithCSLInfo = "";
+
         // Get file names within zip file
-        filesInZip = Object.keys(zip.files);
-        
+        var filesInZip = Object.keys(zip.files);
+
+        // Naive approach: check for one of the expected files
```

Comment:
The current Reference Extractor simply assumes the file it gets to be a .docx (as long as it can be unzipped and it contains one of the `filesToExtract`, it will work). The proposed changes are in line with this simple approach, but it would be possible to add more formal checks if it is useful.